File: src/backend/app/models/OAuthState.java
package models;

import jakarta.persistence.*;

import java.time.Instant;

@Entity
@Table(name = "oauth_states")
public class OAuthState {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Column(name = "state_token", nullable = false, unique = true)
    private String stateToken;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "provider_id", nullable = false)
    private IdentityProvider provider;

    @Column(length = 255)
    private String nonce;

    @Column(name = "redirect_uri", length = 500)
    private String redirectUri;

    @Column(name = "code_verifier", length = 255)
    private String codeVerifier; // For PKCE

    @Column(name = "created_at", nullable = false, updatable = false)
    private Instant createdAt;

    @Column(name = "expires_at", nullable = false)
    private Instant expiresAt;

    // Constructors
    public OAuthState() {}

    public OAuthState(String stateToken, IdentityProvider provider) {
        this.stateToken = stateToken;
        this.provider = provider;
        this.createdAt = Instant.now();
        this.expiresAt = Instant.now().plusSeconds(600); // 10 minutes expiry
    }

    // Lifecycle methods
    @PrePersist
    protected void onCreate() {
        if (createdAt == null) {
            createdAt = Instant.now();
        }
        if (expiresAt == null) {
            expiresAt = Instant.now().plusSeconds(600); // 10 minutes expiry
        }
    }

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getStateToken() {
        return stateToken;
    }

    public void setStateToken(String stateToken) {
        this.stateToken = stateToken;
    }

    public IdentityProvider getProvider() {
        return provider;
    }

    public void setProvider(IdentityProvider provider) {
        this.provider = provider;
    }

    public String getNonce() {
        return nonce;
    }

    public void setNonce(String nonce) {
        this.nonce = nonce;
    }

    public String getRedirectUri() {
        return redirectUri;
    }

    public void setRedirectUri(String redirectUri) {
        this.redirectUri = redirectUri;
    }

    public String getCodeVerifier() {
        return codeVerifier;
    }

    public void setCodeVerifier(String codeVerifier) {
        this.codeVerifier = codeVerifier;
    }

    public Instant getCreatedAt() {
        return createdAt;
    }

    public Instant getExpiresAt() {
        return expiresAt;
    }

    public void setExpiresAt(Instant expiresAt) {
        this.expiresAt = expiresAt;
    }

    // Helper methods
    public boolean isExpired() {
        return Instant.now().isAfter(expiresAt);
    }

    public boolean isValid() {
        return !isExpired();
    }
}

File: src/backendng/src/main/kotlin/com/secman/domain/OAuthState.kt
package com.secman.domain

import jakarta.persistence.*
import java.time.LocalDateTime

@Entity
@Table(name = "oauth_states")
data class OAuthState(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Column(name = "state_value", nullable = false, unique = true)
    var stateValue: String = "",

    @Column(name = "provider_id", nullable = false)
    var providerId: Long = 0,

    @Column(name = "redirect_uri")
    var redirectUri: String? = null,

    @Column(name = "created_at", nullable = false)
    var createdAt: LocalDateTime = LocalDateTime.now(),

    @Column(name = "expires_at", nullable = false)
    var expiresAt: LocalDateTime = LocalDateTime.now().plusMinutes(10) // 10 minute expiry
) {
    @PrePersist
    fun onCreate() {
        createdAt = LocalDateTime.now()
        if (expiresAt.isEqual(LocalDateTime.now()) || expiresAt.isBefore(LocalDateTime.now())) {
            expiresAt = LocalDateTime.now().plusMinutes(10)
        }
    }
}